stdlib/src/memory/boxed.mojo

```diff
@@ -0,0 +1,143 @@
+
+    fn __init__[
+        T: Movable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], owned value: T):
+        """Construct a new Boxed[] by moving the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store. It is restricted to `Movable` here to allow efficient move construction.
+
+        Args:
+            value: The value to move into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, AddressSpace.GENERIC].alloc(1)
+        self._inner.init_pointee_move(value^)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](inout self: Boxed[T, AddressSpace.GENERIC], *, copy_value: T):
+        """Construct a new Boxed[] by explicitly copying the passed value into a new backing allocation.
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_value: The value to explicitly copy into the Boxed[].
+        """
+        self._inner = UnsafePointer[T, address_space].alloc(1)
+        self._inner.init_pointee_explicit_copy(copy_value)
+
+    fn __init__[
+        T: ExplicitlyCopyable
+    ](
+        inout self: Boxed[T, AddressSpace.GENERIC],
+        *,
+        copy_box: Boxed[T, AddressSpace.GENERIC],
+    ):
+        """Construct a new Boxed[] by explicitly copying the value from another Boxed[].
+
+        Parameters:
+            T: The type of the data to store.
+
+        Args:
+            copy_box: The Boxed[] to copy.
+        """
+        self.__init__(copy_value=copy_box[])
+
+    fn __moveinit__(inout self, owned existing: Self):
+        """Move this Boxed[].
+
+        Args:
+            existing: The value to move.
+        """
+        self._inner = existing._inner
+        existing._inner = UnsafePointer[T, address_space]()
+
+    fn __getitem__(
+        ref [_, address_space._value.value]self
+    ) -> ref [__lifetime_of(self._inner), address_space._value.value] T:
+        """Returns a reference to the box's underlying data with parametric mutability.
+
+        Returns:
+            A reference to the data underlying the Boxed[].
+        """
+        # This should have a widening conversion here that allows
+        # the mutable ref that is always (potentially unsafely)
+        # returned from UnsafePointer to be guarded behind the
+        # aliasing guarantees of the lifetime system here.
+        # All of the magic happens above in the function signature
+        var inner_not_null = self._inner.__bool__()
+
+        debug_assert(
+            inner_not_null,
+            (
+                "Box is horribly broken, and __getitem__ was called on a"
+                " destroyed box"
+            ),
+        )
+
+        return self._inner[]
+
+    fn __del__(owned self: Boxed[T, AddressSpace.GENERIC]):
+        """Destroy the Boxed[]."""
+        # check that inner is non-null to accomodate take() and other
+        # consuming end states
+        if self._inner:
```

Comment:
**Suggestion** Remove this runtime branch - see my comment in the `take` method about using `lit.ownership.mark_destroyed`.